blogs/2020/02/14/optimizing-ci.md

```diff
@@ -0,0 +1,117 @@
+Author: Ethan Dennis, João Moreno
+---
+# Improving CI Build Times
+
+February 14, 2020 by Ethan Dennis, [@erdennis13](https://twitter.com/erdennis13) and João Moreno, [@joaomoreno](https://twitter.com/joaomoreno)
+
+Visual Studio Code is a large project with lots of moving parts and an active participant list. We have [shown](https://code.visualstudio.com/blogs/2018/09/12/engineering-with-azure-pipelines) how we actively use Azure Pipelines to keep up with good engineering practices by maintaining our build and continuous integration infrastructure. In this blog post, we’ll talk about how we used the [Azure Pipelines Artifact Caching Tasks](https://github.com/Microsoft/azure-pipelines-artifact-caching-tasks) to dramatically reduce our CI build times.
+
+We described in an [earlier blog post](https://medium.com/crawl-walk-sprint/reducing-vs-code-ci-build-times-by-33-dbb1715b5028) how we reduced CI build times by 33%. This was accomplished by using custom build tasks that cache the node modules that VS Code consumes instead of resolving the packages at build time. While we were happy with this performance boost, we wanted to see how much further we could push the caching tasks that we built.
+
+When we last talked about our CI engineering our target platforms spanned Windows, macOS, and Linux. As of today, VS Code targets a much more diverse set of platforms, such as ARM64 and Alpine Linux for its remote server components. In total, we have eight different targets that all share common build steps. This post outlines how we leveraged the caching tasks to reduce CI duplication and further improve our build times.
+
+## Room for Improvement
+
+So, what exactly were the common steps across all build jobs? Each build target has a job that follows a similar set of steps. At a very high level, each job must:
+
+1. Restore Node packages
+2. Lint the code
+3. Compile the code
+4. Run unit tests
+5. Run integration tests
+
+Our caching tasks were an obvious choice to speed up the “Restore Node packages” step. E.g. why run an expensive `npm install` step, when you could cache the results from a previous run, given that the `package-lock.json` file rarely changes? Since we’ve discussed caching packages previously, what makes this post interesting is how we applied caching to the other steps.
+
+Since linting and compilation are platform-independent, those steps could easily be run by a single build agent that would share its results with other platform-dependent agents. We created a Linux build agent whose sole responsibility was exactly this: restore packages, lint, and compile the code. All we had to do was to share the results with other agents.
+
+## Cache Everything
+
+In order to share cache results across build agents, we needed platform independent caches, which were initially unsupported by the caching tasks. We added a `platformIndependent` parameter to the task for creating and restoring platform-independent caches. For example:
```

Comment:
A general impression I had after reading the blog was that it was not fully clear what is VS Code specific and what is a general pipeline improvement that everyone could leverage.

To make this more clear I suggest the following change:

>We added a `platformIndependent` parameter to the task for creating and restoring platform-independent caches

Azure Pipline task now supports an addition parameter `platformIndependent`. This parameter is leveraged in the VS Code build as follows: